tests/ui/redundant_slicing.stderr

```diff
@@ -1,34 +1,46 @@
 error: redundant slicing of the whole range
-  --> $DIR/redundant_slicing.rs:6:13
+  --> $DIR/redundant_slicing.rs:10:13
    |
-LL |     let _ = &slice[..];
+LL |     let _ = &slice[..]; // Redundant slice
    |             ^^^^^^^^^^ help: use the original value instead: `slice`
    |
    = note: `-D clippy::redundant-slicing` implied by `-D warnings`
 
 error: redundant slicing of the whole range
-  --> $DIR/redundant_slicing.rs:10:13
+  --> $DIR/redundant_slicing.rs:14:13
    |
-LL |     let _ = &(&v[..])[..]; // Outer borrow is redundant
-   |             ^^^^^^^^^^^^^ help: use the original value instead: `(&v[..])`
+LL |     let _ = &(&*v)[..]; // Outer borrow is redundant
+   |             ^^^^^^^^^^ help: use the original value instead: `(&*v)`
```

Comment:
It's been a while since I looked at this PR. I believe if a `*` is suggested, it should be a `deref_by_slicing` case?

Reply:
Apparrently I can't tell the difference between the commit diff and the rust output diff 🤦. But what about the error at the bottom of this file (same question)?

Reply:
I'll assume you're referring to `&slice_ref[..]`. `slice_ref` is a `&&[_]`. Full slicing is the same as just removing a reference in this case.

Reply:
Below that is

```text
error: redundant slicing of the whole range
  --> $DIR/redundant_slicing.rs:45:13
   |
LL |     let _ = (&bytes[..]).read_to_end(&mut vec![]).unwrap();
   |             ^^^^^^^^^^^^ help: reborrow the original value instead: `(&*bytes)`
```

Reply:
`read_to_end` is from an impl on `&[u8]` take `&mut self`. `bytes` is a `&[u8]`. `&bytes[..]` creates a temporary reference to pass to `read_to_end` rather than using the original value. This is essentially slicing in order to re-borrow the reference.

Reply:
Yeah that one seems more like `deref_by_slicing`. Either way it is deref and the borrow, using slicing or deref operator.

Reply:
The reborrow isn't strictly needed. `{bytes}.read_to_end(..)` would also work. I just figured that would end being considered more confusing.

Reply:
If the suggestion were to not reborrow then that would be different. But the suggestion is to replace deref by slicing with deref by `*`. I agree `{bytes}` is not good.

Reply:
If anything the previous one is much more of a deref-by-slicing. This is very much a slice-to-get-a-new-value-so-we-don't-modify-the-current-one (really flows of the tongue there).

I think leaving `redundant_slicing` to only handle cases where it's doing nothing and can be fully removed is fine, but `deref_by_slicing` is not really a good spot to put the remaining cases. That would mean having a name for a new lint (as precise as the name earlier is, I don't want to fight for longest lint name).

Reply:
You could call it slice-to-deref-to-get-a-new-borrow-of... You can put a lot of new names on it but ultimately it just changes `&[u8]` to `[u8]` before the borrow operator applies. I don't see how it is not a deref. The suggestion literally replaces the slicing with a deref operator which has the exact same effect. If I enabled `deref_by_slicing` in my code, I'd consider it a bug if this case were not included.

> I think leaving redundant_slicing to only handle cases where it's doing nothing and can be fully removed is fine

That is the most important thing since it is warn-by-default.

Reply:
We're talking about two slightly different things here. I'm talking about the entire expression `&x[..]` not just `x[,,]`. `x[..]` is just `*x.index(..)` so it will almost always be a deref. In this case the deref part is the only effect being used, so I guess that fits under `deref_by_slicing`.

I'm not sure why you don't take issue with the `&&T` -> `&T` case being under `redundant_slicing`.

Reply:
I'll move that all to `deref_as_slice` then. I still think the re-borrowing case is different enough to warrant it's own lint, but that can be dealt with at another time.